File: typegraph/python_next/typegraph_next/runtimes/python.py
# Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0
import ast
import hashlib
import inspect
from astunparse import unparse

from dataclasses import dataclass
from typing import TYPE_CHECKING, List, Optional

from typegraph_next.runtimes.base import Materializer, Runtime

from typegraph_next.gen.exports.runtimes import (
    Effect,
    EffectNone,
    BaseMaterializer,
    MaterializerPythonImport,
    MaterializerPythonLambdaOrDef,
    MaterializerPythonModule,
)
from typegraph_next.gen.types import Err
from typegraph_next.wit import runtimes, store

if TYPE_CHECKING:
    from typegraph_next import t


class PythonRuntime(Runtime):
    _id: int

    def __init__(self):
        super().__init__(runtimes.register_python_runtime(store))

    def from_lambda(
        self,
        function: callable,
    ):
        effect = EffectNone()
        lambdas, _defs = DefinitionCollector.collect(function)
        assert len(lambdas) == 1
        fn = str(lambdas[0])
        m = hashlib.sha256()
        m.update(fn.encode("utf-8"))
        name = m.hexdigest()

        mat_id = runtimes.from_python_lambda(
            store,
            BaseMaterializer(runtime=self.id.value, effect=effect),
            MaterializerPythonLambdaOrDef(runtime=self.id.value, name=name, fn=fn),
        )

        if isinstance(mat_id, Err):
            raise Exception(mat_id.value)

        return LambdaMat(id=mat_id.value, name=name, fn=fn, effect=effect)

    def from_def(
        self,
        function: callable,
    ):
        effect = EffectNone()
        _lambdas, defs = DefinitionCollector.collect(function)
        assert len(defs) == 1
        name, fn = defs[0]

        mat_id = runtimes.from_python_def(
            store,
            BaseMaterializer(runtime=self.id.value, effect=effect),
            MaterializerPythonLambdaOrDef(runtime=self.id.value, name=name, fn=fn),
        )

        if isinstance(mat_id, Err):
            raise Exception(mat_id.value)

        return DefMat(id=mat_id.value, name=name, fn=fn, effect=effect)

    def import_(
        self,
        inp: "t.struct",
        out: "t.typedef",
        *,
        module: str,
        name: str,
        effect: Optional[Effect] = None,
        secrets: Optional[List[str]] = None,
    ):
        effect = effect or EffectNone()
        secrets = secrets or []

        base = BaseMaterializer(runtime=self.id.value, effect=effect)
        mat_id = runtimes.from_python_module(
            store, base, MaterializerPythonModule(file=module, runtime=self.id.value)
        )

        if isinstance(mat_id, Err):
            raise Exception(mat_id.value)

        py_mod_mat_id = runtimes.from_python_import(
            store,
            base,
            MaterializerPythonImport(
                module=mat_id.value, func_name=name, secrets=secrets
            ),
        )

        if isinstance(py_mod_mat_id, Err):
            raise Exception(py_mod_mat_id.value)

        from typegraph_next import t

        return t.func(
            inp,
            out,
            ImportMat(
                id=py_mod_mat_id.value,
                name=name,
                module=module,
                secrets=secrets,
                effect=effect,
            ),
        )


@dataclass
class LambdaMat(Materializer):
    fn: str
    name: str
    effect: Effect


@dataclass
class DefMat(Materializer):
    fn: str
    name: str
    effect: Effect


class DefinitionCollector(ast.NodeTransformer):
    @classmethod
    def collect(cls, function):
        source = inspect.getsource(function).lstrip()
        tree = ast.parse(source)
        ret = cls()
        ret.visit(tree)
        return ret.lambdas, ret.defs

    def __init__(self):
        super().__init__()
        self.lambdas = []
        self.defs = []

    def visit_Lambda(self, node):
        self.lambdas.append(unparse(node).strip())

    def visit_FunctionDef(self, node):
        self.defs.append((node.name, unparse(node).strip()))


@dataclass
class ImportMat(Materializer):
    module: str
    name: str
    secrets: List[str]

File: typegraph/node/src/runtimes/python.ts
// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

import * as t from "../types.js";
import { runtimes } from "../wit.js";
import { Effect } from "../gen/interfaces/metatype-typegraph-runtimes.js";
import { Materializer, Runtime } from "./mod.js";
import { fx } from "../mod.js";

interface LambdaMat extends Materializer {
  fn: string;
  effect: Effect;
}

interface DefMat extends Materializer {
  fn: string;
  name: string;
  effect: Effect;
}

interface PythonImport {
  name: string;
  module: string;
  secrets?: Array<string>;
  effect?: Effect;
}

interface ImportMat extends Materializer {
  module: string;
  name: string;
}

export class PythonRuntime extends Runtime {
  constructor() {
    super(runtimes.registerPythonRuntime());
  }
  fromLambda<
    P extends Record<string, t.Typedef> = Record<string, t.Typedef>,
    I extends t.Struct<P> = t.Struct<P>,
    O extends t.Typedef = t.Typedef,
  >(inp: I, out: O, { code }: { code: string }) {
    const matId = runtimes.fromPythonLambda(
      {
        runtime: this._id,
        effect: fx.read(),
      },
      {
        fn: code, // not formatted
        runtime: this._id,
      },
    );

    return t.func(inp, out, {
      _id: matId,
      fn: code,
    } as LambdaMat);
  }

  fromDef<
    P extends Record<string, t.Typedef> = Record<string, t.Typedef>,
    I extends t.Struct<P> = t.Struct<P>,
    O extends t.Typedef = t.Typedef,
  >(inp: I, out: O, { code }: { code: string }) {
    const name = code.trim().match(/def\s+([A-Za-z0-9_]+)/)?.[1];
    if (name == undefined) {
      throw new Error(`unable to extract def name from source code ${code}`);
    }
    const matId = runtimes.fromPythonDef(
      {
        runtime: this._id,
        effect: fx.read(),
      },
      {
        name: name,
        fn: code,
        runtime: this._id,
      },
    );

    return t.func(inp, out, {
      _id: matId,
      name,
      fn: code,
    } as DefMat);
  }

  import<
    P extends Record<string, t.Typedef> = Record<string, t.Typedef>,
    I extends t.Struct<P> = t.Struct<P>,
    O extends t.Typedef = t.Typedef,
  >(
    inp: I,
    out: O,
    { name, module, effect = fx.read(), secrets = [] }: PythonImport,
  ): t.Func<P, I, O, ImportMat> {
    const base = {
      runtime: this._id,
      effect,
    };

    const matId = runtimes.fromPythonModule(base, {
      file: module,
      runtime: this._id,
    });

    const pyModMatId = runtimes.fromPythonImport(base, {
      module: matId,
      funcName: name,
      secrets,
    });

    return t.func(inp, out, {
      _id: pyModMatId,
      module,
      name,
    });
  }
}
